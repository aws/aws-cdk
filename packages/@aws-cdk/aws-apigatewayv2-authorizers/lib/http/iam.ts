import {
  HttpAuthorizerType,
  HttpRouteAuthorizerBindOptions,
  HttpRouteAuthorizerConfig,
  IHttpRouteAuthorizer,
} from '@aws-cdk/aws-apigatewayv2';

/**
 * Authorize HTTP API Routes with IAM
 */
export class HttpIamAuthorizer implements IHttpRouteAuthorizer {
  public bind(_options: HttpRouteAuthorizerBindOptions): HttpRouteAuthorizerConfig {
    return {
      authorizationType: HttpAuthorizerType.IAM,
    };
  }
}