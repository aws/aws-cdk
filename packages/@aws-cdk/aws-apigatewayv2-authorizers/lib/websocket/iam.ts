import {
  WebSocketAuthorizerType,
  WebSocketRouteAuthorizerBindOptions,
  WebSocketRouteAuthorizerConfig,
  IWebSocketRouteAuthorizer,
} from '@aws-cdk/aws-apigatewayv2';

/**
 * Authorize WebSocket API Routes with IAM
 */
export class WebSocketIamAuthorizer implements IWebSocketRouteAuthorizer {
  public bind(
    _options: WebSocketRouteAuthorizerBindOptions,
  ): WebSocketRouteAuthorizerConfig {
    return {
      authorizationType: WebSocketAuthorizerType.IAM,
    };
  }
}
