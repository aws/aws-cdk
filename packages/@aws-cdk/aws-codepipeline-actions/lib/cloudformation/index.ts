export * from './pipeline-actions';
export * from './stackset-action';
export * from './stackinstances-action';
export * from './stackset-types';
