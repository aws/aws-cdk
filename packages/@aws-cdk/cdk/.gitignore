*.js
*.d.ts
tsconfig.json
tslint.json
*.js.map
dist
coverage
.nyc_output
.jsii

.LAST_BUILD
.nycrc
.LAST_PACKAGE
*.snk