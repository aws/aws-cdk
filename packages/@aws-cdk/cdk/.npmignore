
dist
.LAST_PACKAGE
*.tsbuildinfo
.LAST_BUILD
*.snk
*.ts
!*.d.ts
!*.js
coverage
.nyc_output
*.tgz

# Include .jsii
!.jsii
