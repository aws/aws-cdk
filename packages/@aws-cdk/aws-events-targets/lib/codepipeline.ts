import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import { singletonEventRole } from './util';

/**
 * Customization options when creating a {@link CodePipeline} event target.
 */
export interface CodePipelineTargetOptions {
  /**
   * The role to assume before invoking the target
   * (i.e., the pipeline) when the given rule is triggered.
   *
   * @default - a new role will be created
   */
  readonly eventRole?: iam.IRole;
}

/**
 * Allows the pipeline to be used as an EventBridge rule target.
 */
export class CodePipeline implements events.IRuleTarget {
  constructor(
    private readonly pipeline: codepipeline.IPipeline,
    private readonly options: CodePipelineTargetOptions = {}) {
  }

  public bind(_rule: events.IRule, _id?: string): events.RuleTargetConfig {
    return {
      id: '',
      arn: this.pipeline.pipelineArn,
      role: this.options.eventRole || singletonEventRole(this.pipeline, [new iam.PolicyStatement({
        resources: [this.pipeline.pipelineArn],
        actions: ['codepipeline:StartPipelineExecution'],
      })]),
      targetResource: this.pipeline,
    };
  }
}
