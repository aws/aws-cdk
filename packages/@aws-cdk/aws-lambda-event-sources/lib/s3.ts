import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import * as notifs from '@aws-cdk/aws-s3-notifications';

export interface S3EventSourceProps {
  /**
   * The s3 event types that will trigger the notification.
   */
  readonly events: s3.EventType[];

  /**
   * S3 object key filter rules to determine which objects trigger this event.
   * Each filter must include a `prefix` and/or `suffix` that will be matched
   * against the s3 object key. Refer to the S3 Developer Guide for details
   * about allowed filter rules.
   */
  readonly filters?: s3.NotificationKeyFilter[];
}

/**
 * Use S3 bucket notifications as an event source for AWS Lambda.
 */
export class S3EventSource implements lambda.IEventSource {
  constructor(readonly bucket: s3.Bucket, private readonly props: S3EventSourceProps) {

  }

  public bind(target: lambda.IFunction) {
    const filters = this.props.filters || [];
    for (const event of this.props.events) {
      this.bucket.addEventNotification(event, new notifs.LambdaDestination(target), ...filters);
    }
  }
}
