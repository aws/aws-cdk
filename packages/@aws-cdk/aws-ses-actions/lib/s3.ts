import * as iam from '@aws-cdk/aws-iam';
import * as kms from '@aws-cdk/aws-kms';
import * as s3 from '@aws-cdk/aws-s3';
import * as ses from '@aws-cdk/aws-ses';
import * as sns from '@aws-cdk/aws-sns';
import * as cdk from '@aws-cdk/core';

/**
 * Construction properties for a S3 action.
 */
export interface S3Props {
  /**
   * The S3 bucket that incoming email will be saved to.
   */
  readonly bucket: s3.IBucket;

  /**
   * The master key that SES should use to encrypt your emails before saving
   * them to the S3 bucket.
   *
   * @default no encryption
   */
  readonly kmsKey?: kms.IKey;

  /**
   * The key prefix of the S3 bucket.
   *
   * @default no prefix
   */
  readonly objectKeyPrefix?: string;

  /**
   * The SNS topic to notify when the S3 action is taken.
   *
   * @default no notification
   */
  readonly topic?: sns.ITopic;
}

/**
 * Saves the received message to an Amazon S3 bucket and, optionally, publishes
 * a notification to Amazon SNS.
 */
export class S3 implements ses.IReceiptRuleAction {
  constructor(private readonly props: S3Props) {
  }

  public bind(rule: ses.IReceiptRule): ses.ReceiptRuleActionConfig {
    // Allow SES to write to S3 bucket
    // See https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-permissions.html#receiving-email-permissions-s3
    const keyPattern = this.props.objectKeyPrefix || '';
    const s3Statement = new iam.PolicyStatement({
      actions: ['s3:PutObject'],
      principals: [new iam.ServicePrincipal('ses.amazonaws.com')],
      resources: [this.props.bucket.arnForObjects(`${keyPattern}*`)],
      conditions: {
        StringEquals: {
          'aws:Referer': cdk.Aws.ACCOUNT_ID,
        },
      },
    });
    this.props.bucket.addToResourcePolicy(s3Statement);

    const policy = this.props.bucket.node.tryFindChild('Policy') as s3.BucketPolicy;
    if (policy) { // The bucket could be imported
      rule.node.addDependency(policy);
    } else {
      rule.node.addWarning('This rule is using a S3 action with an imported bucket. Ensure permission is given to SES to write to that bucket.');
    }

    // Allow SES to use KMS master key
    // See https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-permissions.html#receiving-email-permissions-kms
    if (this.props.kmsKey && !/alias\/aws\/ses$/.test(this.props.kmsKey.keyArn)) {
      const kmsStatement = new iam.PolicyStatement({
        actions: ['km:Encrypt', 'kms:GenerateDataKey'],
        principals: [ new iam.ServicePrincipal('ses.amazonaws.com')],
        resources: ['*'],
        conditions: {
          Null: {
            'kms:EncryptionContext:aws:ses:rule-name': 'false',
            'kms:EncryptionContext:aws:ses:message-id': 'false',
          },
          StringEquals: {
            'kms:EncryptionContext:aws:ses:source-account': cdk.Aws.ACCOUNT_ID,
          },
        },
      });

      this.props.kmsKey.addToResourcePolicy(kmsStatement);
    }

    return {
      s3Action: {
        bucketName: this.props.bucket.bucketName,
        kmsKeyArn: this.props.kmsKey ? this.props.kmsKey.keyArn : undefined,
        objectKeyPrefix: this.props.objectKeyPrefix,
        topicArn: this.props.topic ? this.props.topic.topicArn : undefined,
      },
    };
  }
}
