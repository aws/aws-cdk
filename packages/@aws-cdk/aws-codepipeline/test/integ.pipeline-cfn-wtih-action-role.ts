import cloudformation = require('@aws-cdk/aws-cloudformation');
import iam = require('@aws-cdk/aws-iam');
import s3 = require('@aws-cdk/aws-s3');
import cdk = require('@aws-cdk/cdk');
import codepipeline = require('../lib');

const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-codepipeline-cloudformation-cross-region-with-action-role', {});

const bucket = new s3.Bucket(stack, 'MyBucket', {
  versioned: true,
  removalPolicy: cdk.RemovalPolicy.Destroy,
});

const sourceAction = bucket.toCodePipelineSourceAction({
  actionName: 'S3',
  bucketKey: 'some/path',
});
const sourceStage = {
  name: 'Source',
  actions: [sourceAction],
};

const role = new iam.Role(stack, 'ActionRole', {
  assumedBy: new iam.AccountPrincipal(cdk.Aws.accountId)
});
role.addToPolicy(new iam.PolicyStatement()
  .addAction('sqs:*')
  .addAllResources()
);
const cfnStage = {
  name: 'CFN',
  actions: [
    new cloudformation.PipelineCreateUpdateStackAction({
      actionName: 'CFN_Deploy',
      stackName: 'aws-cdk-codepipeline-cross-region-deploy-stack',
      templatePath: sourceAction.outputArtifact.atPath('template.yml'),
      adminPermissions: false,
      role
    }),
  ],
};

const pipeline = new codepipeline.Pipeline(stack, 'MyPipeline', {
  artifactBucket: bucket,
  stages: [
    sourceStage,
    cfnStage,
  ],
});
pipeline.addToRolePolicy(new iam.PolicyStatement()
  .addActions("sts:AssumeRole", "iam:PassRole")
  .addAllResources()
);

app.run();
