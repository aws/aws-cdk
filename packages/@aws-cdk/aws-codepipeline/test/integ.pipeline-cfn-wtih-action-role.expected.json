{
  "Resources": {
    "MyBucketF68F3FF0": {
      "Type": "AWS::S3::Bucket",
      "Properties": {
        "VersioningConfiguration": {
          "Status": "Enabled"
        }
      }
    },
    "MyPipelineRoleC0D47CA4": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MyPipelineRoleDefaultPolicy34F09EFA": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "MyBucketF68F3FF0",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "MyBucketF68F3FF0",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "MyBucketF68F3FF0",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "MyBucketF68F3FF0",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyPipelineCFNCFNDeployRole9CC99B3F",
                  "Arn"
                ]
              }
            },
            {
              "Action": [
                "cloudformation:CreateStack",
                "cloudformation:DescribeStack*",
                "cloudformation:GetStackPolicy",
                "cloudformation:GetTemplate*",
                "cloudformation:SetStackPolicy",
                "cloudformation:UpdateStack",
                "cloudformation:ValidateTemplate"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":cloudformation:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":stack/aws-cdk-codepipeline-cross-region-deploy-stack/*"
                  ]
                ]
              }
            },
            {
              "Action": [
                "sts:AssumeRole",
                "iam:PassRole"
              ],
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "MyPipelineRoleDefaultPolicy34F09EFA",
        "Roles": [
          {
            "Ref": "MyPipelineRoleC0D47CA4"
          }
        ]
      }
    },
    "MyPipelineAED38ECF": {
      "Type": "AWS::CodePipeline::Pipeline",
      "Properties": {
        "RoleArn": {
          "Fn::GetAtt": [
            "MyPipelineRoleC0D47CA4",
            "Arn"
          ]
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "S3",
                  "Version": "1"
                },
                "Configuration": {
                  "S3Bucket": {
                    "Ref": "MyBucketF68F3FF0"
                  },
                  "S3ObjectKey": "some/path"
                },
                "InputArtifacts": [],
                "Name": "S3",
                "OutputArtifacts": [
                  {
                    "Name": "Artifact_S3_awscdkcodepipelinecloudformationcrossregionwithactionroleMyBucketBDA230BF"
                  }
                ],
                "RunOrder": 1
              }
            ],
            "Name": "Source"
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "CloudFormation",
                  "Version": "1"
                },
                "Configuration": {
                  "StackName": "aws-cdk-codepipeline-cross-region-deploy-stack",
                  "ActionMode": "CREATE_UPDATE",
                  "TemplatePath": "Artifact_S3_awscdkcodepipelinecloudformationcrossregionwithactionroleMyBucketBDA230BF::template.yml",
                  "RoleArn": {
                    "Fn::GetAtt": [
                      "MyPipelineCFNCFNDeployRole9CC99B3F",
                      "Arn"
                    ]
                  }
                },
                "InputArtifacts": [
                  {
                    "Name": "Artifact_S3_awscdkcodepipelinecloudformationcrossregionwithactionroleMyBucketBDA230BF"
                  }
                ],
                "Name": "CFN_Deploy",
                "OutputArtifacts": [],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "ActionRole60B0EDF7",
                    "Arn"
                  ]
                },
                "RunOrder": 1
              }
            ],
            "Name": "CFN"
          }
        ],
        "ArtifactStore": {
          "Location": {
            "Ref": "MyBucketF68F3FF0"
          },
          "Type": "S3"
        }
      },
      "DependsOn": [
        "MyPipelineRoleDefaultPolicy34F09EFA",
        "MyPipelineRoleC0D47CA4"
      ]
    },
    "ActionRole60B0EDF7": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "ActionRoleDefaultPolicyCA33BE56": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "ActionRoleDefaultPolicyCA33BE56",
        "Roles": [
          {
            "Ref": "ActionRole60B0EDF7"
          }
        ]
      }
    },
    "MyPipelineCFNCFNDeployRole9CC99B3F": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudformation.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    }
  }
}
